File: crates/nmp-ffi/src/content.rs
//! UniFFI projection of the optional, UI-blind content semantic layer.

use nmp_content::{
    BlockKind, ContentDiagnostic, ContentSyntax, InlineNode, InlineStyle, ReferencePlacement,
    SourceRange,
};
use uniffi::{Enum, Record};

use crate::reference::{target_to_ffi, FfiReferenceTarget};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Enum)]
pub enum FfiContentSyntax {
    PlainText,
    Markdown,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Record)]
pub struct FfiSourceRange {
    pub start: u32,
    pub end: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Enum)]
pub enum FfiBlockKind {
    Paragraph,
    Heading {
        level: u8,
    },
    Quote {
        depth: u8,
    },
    ListItem {
        ordered: bool,
        ordinal: Option<u64>,
        depth: u8,
    },
    Code {
        language: Option<String>,
    },
    ThematicBreak,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Enum)]
pub enum FfiInlineStyle {
    Emphasis,
    Strong,
    Strikethrough,
    Code,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Enum)]
pub enum FfiReferencePlacement {
    Inline,
    Standalone,
}

#[derive(Debug, Clone, PartialEq, Eq, Record)]
pub struct FfiReferenceOccurrence {
    pub id: u64,
    pub original: String,
    pub target: FfiReferenceTarget,
    pub source: FfiSourceRange,
    pub placement: FfiReferencePlacement,
}

#[derive(Debug, Clone, PartialEq, Eq, Enum)]
pub enum FfiInlineNode {
    Text {
        text: String,
        source: FfiSourceRange,
        styles: Vec<FfiInlineStyle>,
    },
    Reference {
        occurrence: FfiReferenceOccurrence,
        styles: Vec<FfiInlineStyle>,
    },
    Hashtag {
        hashtag: String,
        original: String,
        source: FfiSourceRange,
        styles: Vec<FfiInlineStyle>,
    },
    Link {
        destination: String,
        label: String,
        source: FfiSourceRange,
        styles: Vec<FfiInlineStyle>,
    },
    SoftBreak {
        source: FfiSourceRange,
    },
    HardBreak {
        source: FfiSourceRange,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Record)]
pub struct FfiContentBlock {
    pub id: u64,
    pub kind: FfiBlockKind,
    pub source: FfiSourceRange,
    pub inlines: Vec<FfiInlineNode>,
}

#[derive(Debug, Clone, PartialEq, Eq, Enum)]
pub enum FfiContentDiagnostic {
    InputTruncated {
        original_bytes: u64,
        parsed_bytes: u64,
    },
    MalformedReference {
        original: String,
        source: FfiSourceRange,
    },
}

#[derive(Debug, Clone, PartialEq, Eq, Record)]
pub struct FfiContentDocument {
    pub syntax: FfiContentSyntax,
    pub blocks: Vec<FfiContentBlock>,
    pub diagnostics: Vec<FfiContentDiagnostic>,
}

#[uniffi::export]
pub fn parse_nostr_content(content: String, syntax: FfiContentSyntax) -> FfiContentDocument {
    document_to_ffi(nmp_content::parse_content(
        &content,
        syntax_from_ffi(syntax),
    ))
}

fn syntax_from_ffi(value: FfiContentSyntax) -> ContentSyntax {
    match value {
        FfiContentSyntax::PlainText => ContentSyntax::PlainText,
        FfiContentSyntax::Markdown => ContentSyntax::Markdown,
    }
}

fn syntax_to_ffi(value: ContentSyntax) -> FfiContentSyntax {
    match value {
        ContentSyntax::PlainText => FfiContentSyntax::PlainText,
        ContentSyntax::Markdown => FfiContentSyntax::Markdown,
    }
}

fn range_to_ffi(value: SourceRange) -> FfiSourceRange {
    FfiSourceRange {
        start: value.start,
        end: value.end,
    }
}

fn block_kind_to_ffi(value: BlockKind) -> FfiBlockKind {
    match value {
        BlockKind::Paragraph => FfiBlockKind::Paragraph,
        BlockKind::Heading { level } => FfiBlockKind::Heading { level },
        BlockKind::Quote { depth } => FfiBlockKind::Quote { depth },
        BlockKind::ListItem {
            ordered,
            ordinal,
            depth,
        } => FfiBlockKind::ListItem {
            ordered,
            ordinal,
            depth,
        },
        BlockKind::Code { language } => FfiBlockKind::Code { language },
        BlockKind::ThematicBreak => FfiBlockKind::ThematicBreak,
    }
}

fn style_to_ffi(value: InlineStyle) -> FfiInlineStyle {
    match value {
        InlineStyle::Emphasis => FfiInlineStyle::Emphasis,
        InlineStyle::Strong => FfiInlineStyle::Strong,
        InlineStyle::Strikethrough => FfiInlineStyle::Strikethrough,
        InlineStyle::Code => FfiInlineStyle::Code,
    }
}

fn placement_to_ffi(value: ReferencePlacement) -> FfiReferencePlacement {
    match value {
        ReferencePlacement::Inline => FfiReferencePlacement::Inline,
        ReferencePlacement::Standalone => FfiReferencePlacement::Standalone,
    }
}

fn occurrence_to_ffi(value: nmp_content::ReferenceOccurrence) -> FfiReferenceOccurrence {
    FfiReferenceOccurrence {
        id: value.id,
        original: value.original,
        target: target_to_ffi(value.target),
        source: range_to_ffi(value.source),
        placement: placement_to_ffi(value.placement),
    }
}

fn node_to_ffi(value: InlineNode) -> FfiInlineNode {
    match value {
        InlineNode::Text {
            text,
            source,
            styles,
        } => FfiInlineNode::Text {
            text,
            source: range_to_ffi(source),
            styles: styles.into_iter().map(style_to_ffi).collect(),
        },
        InlineNode::Reference { occurrence, styles } => FfiInlineNode::Reference {
            occurrence: occurrence_to_ffi(occurrence),
            styles: styles.into_iter().map(style_to_ffi).collect(),
        },
        InlineNode::Hashtag {
            hashtag,
            original,
            source,
            styles,
        } => FfiInlineNode::Hashtag {
            hashtag,
            original,
            source: range_to_ffi(source),
            styles: styles.into_iter().map(style_to_ffi).collect(),
        },
        InlineNode::Link {
            destination,
            label,
            source,
            styles,
        } => FfiInlineNode::Link {
            destination,
            label,
            source: range_to_ffi(source),
            styles: styles.into_iter().map(style_to_ffi).collect(),
        },
        InlineNode::SoftBreak { source } => FfiInlineNode::SoftBreak {
            source: range_to_ffi(source),
        },
        InlineNode::HardBreak { source } => FfiInlineNode::HardBreak {
            source: range_to_ffi(source),
        },
    }
}

fn diagnostic_to_ffi(value: ContentDiagnostic) -> FfiContentDiagnostic {
    match value {
        ContentDiagnostic::InputTruncated {
            original_bytes,
            parsed_bytes,
        } => FfiContentDiagnostic::InputTruncated {
            original_bytes,
            parsed_bytes,
        },
        ContentDiagnostic::MalformedReference { original, source } => {
            FfiContentDiagnostic::MalformedReference {
                original,
                source: range_to_ffi(source),
            }
        }
    }
}

fn document_to_ffi(value: nmp_content::ContentDocument) -> FfiContentDocument {
    FfiContentDocument {
        syntax: syntax_to_ffi(value.syntax),
        blocks: value
            .blocks
            .into_iter()
            .map(|block| FfiContentBlock {
                id: block.id,
                kind: block_kind_to_ffi(block.kind),
                source: range_to_ffi(block.source),
                inlines: block.inlines.into_iter().map(node_to_ffi).collect(),
            })
            .collect(),
        diagnostics: value
            .diagnostics
            .into_iter()
            .map(diagnostic_to_ffi)
            .collect(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ffi_document_keeps_reference_and_source_range() {
        let source = "hello npub14f8usejl26twx0dhuxjh9cas7keav9vr0v8nvtwtrjqx3vycc76qqh9nsy";
        let document = parse_nostr_content(source.to_string(), FfiContentSyntax::PlainText);
        assert_eq!(document.blocks.len(), 1);
        assert!(document.blocks[0]
            .inlines
            .iter()
            .any(|node| matches!(node, FfiInlineNode::Reference { .. })));
    }
}

File: Packages/NMPKotlin/src/main/kotlin/com/nmp/sdk/Content.kt
package com.nmp.sdk

import uniffi.nmp_ffi.FfiBlockKind
import uniffi.nmp_ffi.FfiContentDiagnostic
import uniffi.nmp_ffi.FfiContentDocument
import uniffi.nmp_ffi.FfiContentSyntax
import uniffi.nmp_ffi.FfiInlineNode
import uniffi.nmp_ffi.FfiInlineStyle
import uniffi.nmp_ffi.FfiReferenceOccurrence
import uniffi.nmp_ffi.FfiReferencePlacement
import uniffi.nmp_ffi.FfiReferenceTarget
import uniffi.nmp_ffi.FfiSourceRange
import uniffi.nmp_ffi.parseNostrContent as ffiParseNostrContent

enum class NostrContentSyntax {
    PlainText,
    Markdown,
    ;

    internal fun toFfi(): FfiContentSyntax =
        when (this) {
            PlainText -> FfiContentSyntax.PLAIN_TEXT
            Markdown -> FfiContentSyntax.MARKDOWN
        }

    companion object {
        internal fun from(ffi: FfiContentSyntax): NostrContentSyntax =
            when (ffi) {
                FfiContentSyntax.PLAIN_TEXT -> PlainText
                FfiContentSyntax.MARKDOWN -> Markdown
            }
    }
}

data class NostrContentSourceRange(val start: UInt, val end: UInt) {
    companion object {
        internal fun from(ffi: FfiSourceRange) = NostrContentSourceRange(ffi.start, ffi.end)
    }
}

/** Semantic authored context, not a catalog of Compose components. */
sealed class NostrContentBlockContext {
    object Paragraph : NostrContentBlockContext()

    data class Heading(val level: UByte) : NostrContentBlockContext()

    data class Quote(val depth: UByte) : NostrContentBlockContext()

    data class ListItem(
        val ordered: Boolean,
        val ordinal: ULong?,
        val depth: UByte,
    ) : NostrContentBlockContext()

    data class Code(val language: String?) : NostrContentBlockContext()

    object ThematicBreak : NostrContentBlockContext()

    companion object {
        internal fun from(ffi: FfiBlockKind): NostrContentBlockContext =
            when (ffi) {
                is FfiBlockKind.Paragraph -> Paragraph
                is FfiBlockKind.Heading -> Heading(ffi.level)
                is FfiBlockKind.Quote -> Quote(ffi.depth)
                is FfiBlockKind.ListItem -> ListItem(ffi.ordered, ffi.ordinal, ffi.depth)
                is FfiBlockKind.Code -> Code(ffi.language)
                is FfiBlockKind.ThematicBreak -> ThematicBreak
            }
    }
}

enum class NostrContentInlineStyle {
    Emphasis,
    Strong,
    Strikethrough,
    Code,
    ;

    companion object {
        internal fun from(ffi: FfiInlineStyle): NostrContentInlineStyle =
            when (ffi) {
                FfiInlineStyle.EMPHASIS -> Emphasis
                FfiInlineStyle.STRONG -> Strong
                FfiInlineStyle.STRIKETHROUGH -> Strikethrough
                FfiInlineStyle.CODE -> Code
            }
    }
}

enum class NostrReferencePlacement {
    Inline,
    Standalone,
    ;

    companion object {
        internal fun from(ffi: FfiReferencePlacement): NostrReferencePlacement =
            when (ffi) {
                FfiReferencePlacement.INLINE -> Inline
                FfiReferencePlacement.STANDALONE -> Standalone
            }
    }
}

sealed class NostrReferenceTarget {
    data class Profile(
        val pubkey: String,
        val relayHints: List<String> = emptyList(),
    ) : NostrReferenceTarget()

    data class Event(
        val id: String,
        val authorHint: String? = null,
        val kindHint: UShort? = null,
        val relayHints: List<String> = emptyList(),
    ) : NostrReferenceTarget()

    data class Address(
        val kind: UShort,
        val author: String,
        val identifier: String,
        val relayHints: List<String> = emptyList(),
    ) : NostrReferenceTarget()

    val key: String
        get() =
            when (this) {
                is Profile -> "profile:$pubkey"
                is Event -> "event:$id"
                is Address -> "address:$kind:$author:$identifier"
            }

    companion object {
        internal fun from(ffi: FfiReferenceTarget): NostrReferenceTarget =
            when (ffi) {
                is FfiReferenceTarget.Profile -> Profile(ffi.pubkey, ffi.relayHints)
                is FfiReferenceTarget.Event ->
                    Event(ffi.id, ffi.authorHint, ffi.kindHint, ffi.relayHints)
                is FfiReferenceTarget.Address ->
                    Address(ffi.kind, ffi.author, ffi.identifier, ffi.relayHints)
            }
    }
}

data class NostrReferenceOccurrence(
    val id: ULong,
    val original: String,
    val target: NostrReferenceTarget,
    val source: NostrContentSourceRange,
    val placement: NostrReferencePlacement,
) {
    companion object {
        internal fun from(ffi: FfiReferenceOccurrence) =
            NostrReferenceOccurrence(
                ffi.id,
                ffi.original,
                NostrReferenceTarget.from(ffi.target),
                NostrContentSourceRange.from(ffi.source),
                NostrReferencePlacement.from(ffi.placement),
            )
    }
}

sealed class NostrContentInline {
    abstract val source: NostrContentSourceRange

    data class Text(
        val text: String,
        override val source: NostrContentSourceRange,
        val styles: List<NostrContentInlineStyle>,
    ) : NostrContentInline()

    data class Reference(
        val occurrence: NostrReferenceOccurrence,
        val styles: List<NostrContentInlineStyle>,
    ) : NostrContentInline() {
        override val source: NostrContentSourceRange = occurrence.source
    }

    data class Hashtag(
        val hashtag: String,
        val original: String,
        override val source: NostrContentSourceRange,
        val styles: List<NostrContentInlineStyle>,
    ) : NostrContentInline()

    data class Link(
        val destination: String,
        val label: String,
        override val source: NostrContentSourceRange,
        val styles: List<NostrContentInlineStyle>,
    ) : NostrContentInline()

    data class SoftBreak(override val source: NostrContentSourceRange) : NostrContentInline()

    data class HardBreak(override val source: NostrContentSourceRange) : NostrContentInline()

    companion object {
        internal fun from(ffi: FfiInlineNode): NostrContentInline =
            when (ffi) {
                is FfiInlineNode.Text ->
                    Text(
                        ffi.text,
                        NostrContentSourceRange.from(ffi.source),
                        ffi.styles.map(NostrContentInlineStyle::from),
                    )
                is FfiInlineNode.Reference ->
                    Reference(
                        NostrReferenceOccurrence.from(ffi.occurrence),
                        ffi.styles.map(NostrContentInlineStyle::from),
                    )
                is FfiInlineNode.Hashtag ->
                    Hashtag(
                        ffi.hashtag,
                        ffi.original,
                        NostrContentSourceRange.from(ffi.source),
                        ffi.styles.map(NostrContentInlineStyle::from),
                    )
                is FfiInlineNode.Link ->
                    Link(
                        ffi.destination,
                        ffi.label,
                        NostrContentSourceRange.from(ffi.source),
                        ffi.styles.map(NostrContentInlineStyle::from),
                    )
                is FfiInlineNode.SoftBreak ->
                    SoftBreak(NostrContentSourceRange.from(ffi.source))
                is FfiInlineNode.HardBreak ->
                    HardBreak(NostrContentSourceRange.from(ffi.source))
            }
    }
}

data class NostrContentBlock(
    val id: ULong,
    val context: NostrContentBlockContext,
    val source: NostrContentSourceRange,
    val inlines: List<NostrContentInline>,
)

sealed class NostrContentDiagnostic {
    data class InputTruncated(val originalBytes: ULong, val parsedBytes: ULong) :
        NostrContentDiagnostic()

    data class MalformedReference(
        val original: String,
        val source: NostrContentSourceRange,
    ) : NostrContentDiagnostic()

    companion object {
        internal fun from(ffi: FfiContentDiagnostic): NostrContentDiagnostic =
            when (ffi) {
                is FfiContentDiagnostic.InputTruncated ->
                    InputTruncated(ffi.originalBytes, ffi.parsedBytes)
                is FfiContentDiagnostic.MalformedReference ->
                    MalformedReference(ffi.original, NostrContentSourceRange.from(ffi.source))
            }
    }
}

data class NostrContentDocument(
    val syntax: NostrContentSyntax,
    val blocks: List<NostrContentBlock>,
    val diagnostics: List<NostrContentDiagnostic>,
) {
    val references: List<NostrReferenceOccurrence>
        get() =
            blocks.flatMap { it.inlines }.mapNotNull { inline ->
                (inline as? NostrContentInline.Reference)?.occurrence
            }

    companion object {
        internal fun from(ffi: FfiContentDocument) =
            NostrContentDocument(
                syntax = NostrContentSyntax.from(ffi.syntax),
                blocks =
                    ffi.blocks.map { block ->
                        NostrContentBlock(
                            id = block.id,
                            context = NostrContentBlockContext.from(block.kind),
                            source = NostrContentSourceRange.from(block.source),
                            inlines = block.inlines.map(NostrContentInline::from),
                        )
                    },
                diagnostics = ffi.diagnostics.map(NostrContentDiagnostic::from),
            )
    }
}

fun parseNostrContent(
    content: String,
    syntax: NostrContentSyntax = NostrContentSyntax.PlainText,
): NostrContentDocument = NostrContentDocument.from(ffiParseNostrContent(content, syntax.toFfi()))
